### PR TITLE
Continued battle with test_distributed

### PR DESCRIPTION
**Pull request by stephenroller**
**Patch description**
We're finding its very difficult to control for random seed exactly across all workers, causing models to initialize slightly differently. Previously, we just checked that the norm of the model parameters was the same, and raised an exception when they weren't. This test was failing quite frequently.

Now instead, we just explicitly use the primary worker's copy of the parameters, and force a sync at the beginning of the program. This is much easier than trying to control the random seed, and is relatively cheap compared to the full training run.

OLD MESSAGE: @gchanan and @mrshenli suggested we actually check all the parameters rather than an aggregation, possibly improving the reliability of distributed.

**Testing steps**
We're going to run CircleCI 20 times.

### CODE REVIEW

**Comment by stephenroller** [comment]
Ready for reviww

File: parlai/core/build_data.py [hunk 1]
```diff
@@ -21,7 +21,11 @@
 import tqdm
 import math
 import zipfile
-from multiprocessing import Pool
+
+try:
+    from torch.multiprocessing import Pool
```
Comment (stephenroller) [inline parlai/core/build_data.py]: using torch.multiprocessing fixes the issues around spawn that I was originally fighting, hence the changes in the import method.

File: parlai/utils/distributed.py [hunk 2]
```diff
@@ -229,21 +230,27 @@ def sync_object(data, max_size=16384):
...
+    Sync all parameters across all workers are the same.
 
-    Always returns True, or raises an AssertionError if they are not
-    synchronized.
+    Always returns True, or raises an AssertionError if there was a failure.
 
-    :param torch.nn.Module model: A pytorch model.
-    :return: True
+    :param model: A pytorch model.
+    :return: always True
```
Comment (klshuster) [inline parlai/utils/distributed.py]: why?

File: parlai/utils/distributed.py [hunk 3]
```diff
@@ -229,21 +230,27 @@ def sync_object(data, max_size=16384):
...
     """
     if not is_distributed():
         # if things aren't distributed, of course things are in sync
         return True
 
-    # compute the local norm:
+    # syncrhonize all the parameters
```
Comment (klshuster) [inline parlai/utils/distributed.py]: nit: synchronize

File: parlai/utils/distributed.py [hunk 4]
```diff
@@ -229,21 +230,27 @@ def sync_object(data, max_size=16384):
...
+    with torch.no_grad():
+        for p in model.parameters():
+            if not is_primary_worker():
+                p.data.zero_()
+            dist.all_reduce(p.data)
```
Comment (klshuster) [inline parlai/utils/distributed.py]: does all_reduce need to be called with primary worker or is it ok for any of the workers to call it? seems like this would make all params 0?

File: parlai/utils/distributed.py (hunk 3 above)
Comment (stephenroller) [reply parlai/utils/distributed.py]: how do i manage to write any code correctly?

File: parlai/utils/distributed.py (hunk 2 above)
Comment (stephenroller) [reply parlai/utils/distributed.py]: it could return None. it just felt like a nice thing to me.

File: parlai/utils/distributed.py (hunk 4 above)
Comment (stephenroller) [reply parlai/utils/distributed.py]: all_reduce sums the result across all workers, including the primary worker. Hence we want to zero everyone out, EXCEPT the primary worker, and then take the sum. The result will be everyone has the primary worker's parameters.

File: parlai/utils/distributed.py (hunk 2 above)
Comment (klshuster) [reply parlai/utils/distributed.py]: ahh i am ok with `True` I just wasn't sure if it was being used

File: parlai/utils/distributed.py (hunk 4 above)
Comment (klshuster) [reply parlai/utils/distributed.py]: ok, thank you, didn't realize the implicit op was sum but that makes sense

Comment (stephenroller) [reply parlai/utils/distributed.py]: it has an optional reduce arg which I'll add for clarity
